File: Graph/785. Is Graph Bipartite?/isGraphBipartite.py
class Solution:
    def isBipartite(self, graph: List[List[int]]) -> bool:
        visited = [False] * 100
        colours = [False] * 100
        
        queue = []
        
        for i in range(len(graph)):
            if len(graph[i]) == 0 or visited[i] == True:
                continue
            
            queue.append(i)
            visited[i] = True
            colours[i] = True

            while len(queue) != 0:
                node = queue.pop(0)
                for nei in graph[node]:
                    if visited[nei] == False:
                        queue.append(nei)
                        # mark nei as visited
                        visited[nei] = True
                        # colour of nei should be the opposite
                        # to the colour of node
                        colours[nei] = not colours[node]
                    elif colours[node] == colours[nei]:
                        return False
        return True

File: Graph/785. Is Graph Bipartite?/isGraphBipartite.js
/**
 * @param {number[][]} graph
 * @return {boolean}
 */
 var isBipartite = function(graph) {
    // 1 <= n <= 100
    let visited = new Array(100).fill(false);
    let colors = new Array(100).fill(false);
    let queue = [];
    
    for (let i = 0; i < graph.length; i++) {
        // only visit what has not been visited
        if (graph[i].length === 0 || visited[i] === true) {
            continue;
        }
        
        // queue contains what'll be visited
        queue.push(i);
        visited[i] = true;
        colors[i] = true;
        
        while (queue.length !== 0) {
            let node = queue.pop();
            
            for (let neighbour of graph[node]) {
                if (visited[neighbour] === false){
                    queue.push(neighbour);
                    visited[neighbour] = true;
                    // toggle colors
                    colors[neighbour] = !colors[node];  
                }
                else if (colors[node] === colors[neighbour]) {
                    return false;
                }
            } 
        }
    }
    
    return true;
};